### DISCUSSION TITLE
top-level disjunction in a definition: a list of strings or a string, and joining if a list

### DISCUSSION THREAD

**Discussion by sigwinch28** [cue-lang/cue#1655] [Q&A] [answered]
I was trying to figure out a way of using a list of strings or a single string as arguments for calling some CLI tool, wherein a list of strings would need joining together with commas.

I'm using dagger, but there's nothing dagger-specific that I'm having a problem with. I think I'm failing to understand how disjunctions work or _should_ be used.

For example, given the string "x", I want the string "mycmd x"
Given the list of strings `["y", "z"]` I want the string "mydmd y, z".

I was trying to use a disjunction like so:

```cue
import ( "strings" )

#Attempt: {
    {
        a: [...string]
        _a: strings.Join(a, ", ")
    } | {
        a: string,
        _a: a
    }
    
    b: "mycmd \(_a)"
}

#Attempt & {
    a: "x"
}
```

but I get:
```
failed to build: -:12:17: reference "_a" not found
```

How do I achieve what I want, which is to have `b` to be "mycmd x"? I don't want to make `_a` visible externally. Is this a case of me needing to access `_a` in a certain way, or something else?

**Comment by sigwinch28** [comment]
I've found a way to achieve this: 

```cue
import ( "strings" )

#Attempt: {
    a: [...string] | string
    _a: string
    { a: [...string], _a: strings.Join(a, ", ") } | { a: string, _a: a }
    
    b: "mycmd \(_a)"
}

{
    one: #Attempt & {
        a: "x"
    }
    two: #Attempt & {
        a: ["y", "z"]
    }
}
```

```cue
import ( "strings" )

#Attempt: ({
    a: string
    _a: a
} | {
    a: [...string]
    _a: strings.Join(a, ", ")
}) & {

    _a: string
    b: "mycmd \(_a)"
} // unification of a disjunction of definitions

{
    one: #Attempt & {
        a: "x"
    }
    two: #Attempt & {
        a: ["y", "z"]
    }
}
```

both of which are concrete and produce the following:
```cue
one: {
	a: "x"
	b: "mycmd x"
}
two: {
	a: ["y", "z"]
	b: "mycmd y, z"
}
```

**Comment by slewiskelly** [comment, marked as answer]
From your first example, the following would have got it to work:
```cue
#Attempt: {
    _a: _

    {
        a: [...string]
        _a: strings.Join(a, ", ")
    } | {
        a: string
        _a: a
    }
    
    b: "mycmd \(_a)"
}
```

However, the most simple way I can think of, would be:

```cue
#Attempt: {
    a: string | [...string] 
    
    _a: [
        if (a & [...string]) != _|_ { strings.Join(a, ", ") },
        a
    ][0]
    
    
    b: "mycmd \(_a)"
}
```
